stdlib/dynarray.mli

```diff
@@ -0,0 +1,190 @@
+(** Init the array with the given function and size. *)
+
+val clear : 'a t -> unit
+(** Clear the content of the array.
+    This ensures that [length v = 0] but the underlying array is kept,
+    and possibly references to former elements, which are therefore
+    not garbage collectible. *)
+
+val ensure_capacity_with : 'a t -> filler:'a -> int -> unit
+(** Make sure that the array has at least the given capacity (underlying size).
+
+    This is a more advanced operation that is only useful for performance
+    purposes.
+
+    @param filler an element used if the underlying array is empty,
+      to initialize it. It will be retained until the array is totally
+      empty or until it is garbage collected.
+    @raise Invalid_arg if the size is not suitable (negative, or too big for
+    OCaml arrays)
+*)
+
+val ensure_capacity_nonempty : 'a t -> int -> unit
+(** Make sure that the array has at least the given capacity (underlying size),
+    assuming it is non-empty. The first element is used as the filler.
+
+    This is a more advanced operation that is only useful for performance
+    purposes.
+
+    @raise Invalid_arg if the array is empty or
+      if the size is not suitable (negative, or too big for OCaml arrays)
+*)
+
+val is_empty : 'a t -> bool
+(** Is the array empty? This is synonymous to [length a = 0]. *)
+
+val push : 'a t -> 'a -> unit
+(** Add an element at the end of the array. This might extend the underlying
+    array if it is full.
+
+    Calling [push] [n] times is amortized O(n) complexity,
+    and O(ln(n)) reallocations of the underlying array. *)
+
+val unsafe_push : 'a t -> 'a -> unit
+(** Push an element, assuming there is capacity for it
+    (e.g. using {!ensure_capacity}).
+
+    It is unspecified what happens if the capacity is not enough.
+    This is for advanced used only. *)
+
+val append : 'a t -> 'a t -> unit
+(** [append a b] adds all elements of [b] to [a]. [b] is not modified. *)
+
+val append_array : 'a t -> 'a array -> unit
+(** Like {!append}, with an array. *)
+
+val append_seq : 'a t -> 'a Seq.t -> unit
+(** Like {!append} but with an iterator. *)
+
+val append_list : 'a t -> 'a list -> unit
+(** Like {!append} but with a list. *)
+
+val pop : 'a t -> 'a option
+(** Remove and return the last element, or [None] if the
+    array is empty. *)
+
+val pop_exn : 'a t -> 'a
+(** Remove the last element, or raise an exception if the
+    array is empty.
+    @raise Invalid_argument on an empty array. *)
+
+val copy : 'a t -> 'a t
+(** Shallow copy. *)
+
+val truncate : 'a t -> int -> unit
+(** Truncate to the given size (remove elements above this size).
+    Does nothing if the parameter is bigger than the current size.
+
+    [truncate arr n] is similar to:
+    [while length arr > n do ignore (pop_exn arr) done] *)
+
+val shrink_capacity : 'a t -> unit
+(** Shrink internal array to fit the size of the array. This can be useful
+    to make sure there is no memory wasted on a long-held array. *)
```

Comment:
I think this name is confusing. Didn't think long about it but `make_tight`, `fit_capacity`, `tighten_capacity`, `make_capacity_tight` perhaps ?

Reply:
Possibly, `fit_capacity` sounds good I think.